internal/pkg/template/template_functions.go

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package template
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	dashReplacement = "DASH"
+)
+
+// ReplaceDashesFunc takes a CloudFormation logical ID, and
+// sanitizes it by removing "-" characters (not allowed)
+// and replacing them with "DASH" (allowed by CloudFormation but
+// not permitted in ecs-cli generated resource names).
+func ReplaceDashesFunc(logicalID string) string {
+	return strings.ReplaceAll(logicalID, "-", dashReplacement)
+}
+
+// DashReplacedLogicalIDToOriginal takes a "sanitized" logical ID
+// and converts it back to its original form, with dashes.
+func DashReplacedLogicalIDToOriginal(safeLogicalID string) string {
+	return strings.ReplaceAll(safeLogicalID, dashReplacement, "-")
+}
+
+var nonAlphaNum = regexp.MustCompile("[^a-zA-Z0-9]+")
+
+// LogicalIDSafeFunc strips non-alphanumeric characters from an input string.
+func LogicalIDSafeFunc(s string) string {
+	return nonAlphaNum.ReplaceAllString(s, "")
```

Comment:
Do we need to call `ReplaceDashesFunc` to make a string logical id safe?

Reply:
`ReplaceDashesFunc` is used when writing the service cloudformation and has slightly different behavior than the `LogicalIDSafeFunc`, which is used when writing addons templates.

Originally, `ReplaceDashesFunc` was the only LogicalIDSafe method we needed, as we only needed to turn service names into CF logical IDs. Service names can only include lowercase letters, numbers, and dashes, so for example it would turn `my-cool-frontend` into `myDASHcoolDASHfrontend` for the service CF logical IDs which related to that service. 

We also needed to be able to recover the name of the service from the logical ID in methods in [internal/pkg/deploy/cloudformation/stack](https://github.com/aws/amazon-ecs-cli-v2/blob/b29a819367a4400cfa2ea325f0809b7d5871846e/internal/pkg/deploy/cloudformation/stack/app.go#L218) used in calls to `pipeline update` (this method is buried deep inside a call to `GetRegionalAppResources` and part of that is extracting original IDs from logical IDs for ECR repo URL generation).

With the addition of `storage init`, though, we needed to be able to sanitize more complex strings into logical IDs, as DDB table names can include uppercase, lowercase, numbers, and `._-` characters. This is why I added `LogicalIDSafeFunc`, to handle this more complex parsing, as it results in simpler env var names injected into the container. 

Does that make sense?